src/irmin-pack/unix/ranges.ml

```diff
@@ -0,0 +1,125 @@
+(*
+ * Copyright (c) 2022-2023 Tarides <contact@tarides.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open! Import
+
+module Stack = struct
+  type t = Empty | Stack of { mutable len : int; arr : int63 array; prev : t }
+
+  let capacity = 131_072 (* = 128*1024, a large but not too large chunk size *)
+  let make prev = Stack { len = 0; arr = Array.make capacity Int63.zero; prev }
+  let is_full = function Empty -> true | Stack s -> s.len >= capacity
+
+  let rec push_pair ~off ~len t =
+    match t with
+    | Stack s when not (is_full t) ->
+        let i = s.len in
+        s.len <- i + 2;
+        s.arr.(i) <- off;
+        s.arr.(i + 1) <- len;
+        t
+    | _ -> push_pair ~off ~len (make t)
+
+  let rec to_seq t () =
+    match t with
+    | Empty -> Seq.Nil
+    | Stack { len; arr; prev } ->
+        assert (len mod 2 = 0);
+        let rec go i () =
+          if i < 0 then to_seq prev ()
+          else
+            let entry = (arr.(2 * i), arr.((2 * i) + 1)) in
+            Seq.Cons (entry, go (i - 1))
+        in
+        go ((len / 2) - 1) ()
+end
+
+type entry = int63 * int63
```

Comment:
Nit: add doc comment to indicate what each `int63` represents (ie, `(offset, len)`)